File: backend/src/main/java/com/productivity/service/HabitService.java
package com.productivity.service;

import com.google.cloud.Timestamp;
import com.productivity.dto.HabitDTO;
import com.productivity.dto.HabitLogDTO;
import com.productivity.exception.ResourceNotFoundException;
import com.productivity.model.Habit;
import com.productivity.model.HabitLog;
import com.productivity.repository.HabitLogRepository;
import com.productivity.repository.HabitRepository;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;

import java.time.LocalDate;
import java.time.format.DateTimeFormatter;
import java.time.temporal.ChronoUnit;
import java.util.List;
import java.util.Map;
import java.util.HashMap;
import java.util.concurrent.ExecutionException;
import java.util.stream.Collectors;
import com.productivity.dto.HabitDashboardDTO;
import com.productivity.dto.HabitIntelligenceDTO;

@Service
public class HabitService {

    private static final Logger log = LoggerFactory.getLogger(HabitService.class);
    private static final DateTimeFormatter DATE_FORMAT = DateTimeFormatter.ISO_LOCAL_DATE;

    private final HabitRepository habitRepository;
    private final HabitLogRepository habitLogRepository;

    public HabitService(HabitRepository habitRepository, HabitLogRepository habitLogRepository) {
        this.habitRepository = habitRepository;
        this.habitLogRepository = habitLogRepository;
    }

    public HabitDTO createHabit(String userId, HabitDTO dto) throws ExecutionException, InterruptedException {
        Habit habit = new Habit();
        habit.setName(dto.getName());
        habit.setDescription(dto.getDescription());
        habit.setCategory(dto.getCategory());
        habit.setFrequency(dto.getFrequency());
        habit.setGoalType(dto.getGoalType());
        habit.setGoalValue(dto.getGoalValue());
        habit.setMotivation(dto.getMotivation());
        habit.setCurrentStreak(0);
        habit.setLongestStreak(0);
        habit.setCreatedAt(Timestamp.now());

        Habit saved = habitRepository.save(userId, habit);
        return toDTO(saved);
    }

    public List<HabitDTO> getHabits(String userId) throws ExecutionException, InterruptedException {
        return habitRepository.findAll(userId).stream()
                .map(this::toDTO).collect(Collectors.toList());
    }

    public HabitDTO getHabit(String userId, String habitId) throws ExecutionException, InterruptedException {
        Habit habit = habitRepository.findById(userId, habitId)
                .orElseThrow(() -> new ResourceNotFoundException("Habit", habitId));
        return toDTO(habit);
    }

    public HabitDTO updateHabit(String userId, String habitId, HabitDTO dto) throws ExecutionException, InterruptedException {
        Habit existing = habitRepository.findById(userId, habitId)
                .orElseThrow(() -> new ResourceNotFoundException("Habit", habitId));

        if (dto.getName() != null) existing.setName(dto.getName());
        if (dto.getDescription() != null) existing.setDescription(dto.getDescription());
        if (dto.getCategory() != null) existing.setCategory(dto.getCategory());
        if (dto.getFrequency() != null) existing.setFrequency(dto.getFrequency());
        if (dto.getGoalType() != null) existing.setGoalType(dto.getGoalType());
        if (dto.getGoalValue() > 0) existing.setGoalValue(dto.getGoalValue());
        if (dto.getMotivation() != null) existing.setMotivation(dto.getMotivation());

        Habit saved = habitRepository.save(userId, existing);
        return toDTO(saved);
    }

    public void deleteHabit(String userId, String habitId) throws ExecutionException, InterruptedException {
        habitRepository.findById(userId, habitId)
                .orElseThrow(() -> new ResourceNotFoundException("Habit", habitId));
        habitRepository.delete(userId, habitId);
    }

    /**
     * Mark a habit as completed for today. Handles streak logic:
     * - If last completed was yesterday → increment streak
     * - If last completed was today → already done, return
     * - Otherwise → reset streak to 1 (missed day)
     */
    public HabitDTO completeHabit(String userId, String habitId) throws ExecutionException, InterruptedException {
        Habit habit = habitRepository.findById(userId, habitId)
                .orElseThrow(() -> new ResourceNotFoundException("Habit", habitId));

        String today = LocalDate.now().format(DATE_FORMAT);

        // Check if already completed today
        if (today.equals(habit.getLastCompletedDate())) {
            log.info("Habit {} already completed today for user {}", habitId, userId);
            return toDTO(habit);
        }

        // Save the log entry
        HabitLog logEntry = new HabitLog();
        logEntry.setDate(today);
        logEntry.setCompleted(true);
        logEntry.setCompletedAt(Timestamp.now());
        habitLogRepository.save(userId, habitId, logEntry);

        // Calculate streak
        if (habit.getLastCompletedDate() != null) {
            LocalDate lastDate = LocalDate.parse(habit.getLastCompletedDate(), DATE_FORMAT);
            long daysBetween = ChronoUnit.DAYS.between(lastDate, LocalDate.now());

            if (daysBetween == 1) {
                // Consecutive day — increment streak
                habit.setCurrentStreak(habit.getCurrentStreak() + 1);
            } else {
                // Missed day(s) — reset streak
                habit.setCurrentStreak(1);
            }
        } else {
            // First completion
            habit.setCurrentStreak(1);
        }

        // Update longest streak
        if (habit.getCurrentStreak() > habit.getLongestStreak()) {
            habit.setLongestStreak(habit.getCurrentStreak());
        }

        habit.setLastCompletedDate(today);
        Habit saved = habitRepository.save(userId, habit);
        return toDTO(saved);
    }

    public List<HabitLogDTO> getHabitLogs(String userId, String habitId, String startDate, String endDate)
            throws ExecutionException, InterruptedException {
        habitRepository.findById(userId, habitId)
                .orElseThrow(() -> new ResourceNotFoundException("Habit", habitId));

        return habitLogRepository.findByDateRange(userId, habitId, startDate, endDate)
                .stream().map(this::toLogDTO).collect(Collectors.toList());
    }

    public HabitDashboardDTO getDashboard(String userId) throws ExecutionException, InterruptedException {
        List<HabitDTO> habits = getHabits(userId);
        
        String endDate = LocalDate.now().format(DATE_FORMAT);
        String startDate = LocalDate.now().minusDays(30).format(DATE_FORMAT);
        
        Map<String, List<HabitLogDTO>> logs = new HashMap<>();
        for (HabitDTO h : habits) {
            logs.put(h.getId(), getHabitLogs(userId, h.getId(), startDate, endDate));
        }
        
        HabitIntelligenceDTO intelligence = computeIntelligence(habits, logs);
        
        return new HabitDashboardDTO(habits, logs, intelligence);
    }

    private HabitIntelligenceDTO computeIntelligence(List<HabitDTO> habits, Map<String, List<HabitLogDTO>> logsMap) {
        HabitIntelligenceDTO intel = new HabitIntelligenceDTO();
        if (habits.isEmpty()) return intel;
        
        LocalDate today = LocalDate.now();
        
        // 1. Consistency Score (last 30 days weighted)
        double weightedCompletion = 0;
        double totalWeight = 0;
        
        for (int i = 0; i < 30; i++) {
            LocalDate date = today.minusDays(i);
            String dateStr = date.format(DATE_FORMAT);
            double weight = (30.0 - i) / 30.0;
            totalWeight += weight * habits.size();
            
            for (HabitDTO h : habits) {
                List<HabitLogDTO> hLogs = logsMap.get(h.getId());
                if (hLogs != null && hLogs.stream().anyMatch(l -> l.getDate().equals(dateStr) && l.isCompleted())) {
                    weightedCompletion += weight;
                }
            }
        }
        int consistency = totalWeight > 0 ? (int) Math.round((weightedCompletion / totalWeight) * 100) : 0;
        intel.setConsistencyScore(consistency);
        
        // 2. Risk Detection
        int riskCount = 0;
        String todayStr = today.format(DATE_FORMAT);
        String yesterdayStr = today.minusDays(1).format(DATE_FORMAT);
        
        for (HabitDTO h : habits) {
            if ("daily".equals(h.getFrequency()) || h.getFrequency() == null) {
                if (h.getCurrentStreak() > 0 && !todayStr.equals(h.getLastCompletedDate()) && !yesterdayStr.equals(h.getLastCompletedDate())) {
                    riskCount++;
                }
            }
        }
        intel.setRiskCount(riskCount);
        
        // 3. Longest Streak
        int maxStreak = habits.stream().mapToInt(HabitDTO::getCurrentStreak).max().orElse(0);
        intel.setLongestStreak(maxStreak);
        
        // 4. Weekly Completion Rate
        int weeklyCompleted = 0;
        int weeklyExpected = habits.size() * 7;
        for (int i = 0; i < 7; i++) {
            String dateStr = today.minusDays(i).format(DATE_FORMAT);
            for (HabitDTO h : habits) {
                List<HabitLogDTO> hLogs = logsMap.get(h.getId());
                if (hLogs != null && hLogs.stream().anyMatch(l -> l.getDate().equals(dateStr) && l.isCompleted())) {
                    weeklyCompleted++;
                }
            }
        }
        int weeklyRate = weeklyExpected > 0 ? (int) Math.round((weeklyCompleted * 100.0) / weeklyExpected) : 0;
        intel.setWeeklyCompletionRate(weeklyRate);
        
        return intel;
    }

    private HabitDTO toDTO(Habit habit) {
        HabitDTO dto = new HabitDTO();
        dto.setId(habit.getId());
        dto.setName(habit.getName());
        dto.setDescription(habit.getDescription());
        dto.setCategory(habit.getCategory());
        dto.setFrequency(habit.getFrequency());
        dto.setGoalType(habit.getGoalType());
        dto.setGoalValue(habit.getGoalValue());
        dto.setMotivation(habit.getMotivation());
        dto.setCurrentStreak(habit.getCurrentStreak());
        dto.setLongestStreak(habit.getLongestStreak());
        dto.setLastCompletedDate(habit.getLastCompletedDate());
        if (habit.getCreatedAt() != null) {
            dto.setCreatedAt(habit.getCreatedAt().toDate().toInstant().toString());
        }
        return dto;
    }

    private HabitLogDTO toLogDTO(HabitLog log) {
        HabitLogDTO dto = new HabitLogDTO();
        dto.setDate(log.getDate());
        dto.setCompleted(log.isCompleted());
        if (log.getCompletedAt() != null) {
            dto.setCompletedAt(log.getCompletedAt().toDate().toInstant().toString());
        }
        return dto;
    }
}

File: backend-node/src/services/habitService.js
'use strict';

const habitRepo = require('../repositories/habitRepository');
const habitLogRepo = require('../repositories/habitLogRepository');
const { toISO } = require('../utils/timestamp');

function notFound(id) {
  return Object.assign(new Error(`Habit not found: ${id}`), { status: 404 });
}

function todayStr() {
  return new Date().toISOString().slice(0, 10); // YYYY-MM-DD
}

function dateStrMinusDays(n) {
  const d = new Date();
  d.setDate(d.getDate() - n);
  return d.toISOString().slice(0, 10);
}

function daysBetween(dateA, dateB) {
  const msA = new Date(dateA).getTime();
  const msB = new Date(dateB).getTime();
  return Math.round(Math.abs(msB - msA) / 86400000);
}

async function createHabit(userId, dto) {
  const habit = {
    name: dto.name,
    description: dto.description || null,
    category: dto.category || null,
    frequency: dto.frequency || 'daily',
    goalType: dto.goalType || null,
    goalValue: dto.goalValue || 0,
    motivation: dto.motivation || null,
    currentStreak: 0,
    longestStreak: 0,
    lastCompletedDate: null,
    createdAt: new Date(),
  };
  const saved = await habitRepo.save(userId, habit);
  return toDTO(saved);
}

async function getHabits(userId) {
  const habits = await habitRepo.findAll(userId);
  return habits.map(toDTO);
}

async function getHabit(userId, habitId) {
  const habit = await habitRepo.findById(userId, habitId);
  if (!habit) throw notFound(habitId);
  return toDTO(habit);
}

async function updateHabit(userId, habitId, dto) {
  const existing = await habitRepo.findById(userId, habitId);
  if (!existing) throw notFound(habitId);

  if (dto.name != null) existing.name = dto.name;
  if (dto.description != null) existing.description = dto.description;
  if (dto.category != null) existing.category = dto.category;
  if (dto.frequency != null) existing.frequency = dto.frequency;
  if (dto.goalType != null) existing.goalType = dto.goalType;
  if (dto.goalValue != null && dto.goalValue > 0) existing.goalValue = dto.goalValue;
  if (dto.motivation != null) existing.motivation = dto.motivation;

  const saved = await habitRepo.save(userId, existing);
  return toDTO(saved);
}

async function deleteHabit(userId, habitId) {
  const habit = await habitRepo.findById(userId, habitId);
  if (!habit) throw notFound(habitId);
  await habitRepo.remove(userId, habitId);
}

async function completeHabit(userId, habitId) {
  const habit = await habitRepo.findById(userId, habitId);
  if (!habit) throw notFound(habitId);

  const today = todayStr();

  // Idempotent — already done today
  if (habit.lastCompletedDate === today) return toDTO(habit);

  // Persist log
  await habitLogRepo.save(userId, habitId, {
    date: today,
    completed: true,
    completedAt: new Date().toISOString(),
  });

  // Streak logic
  if (habit.lastCompletedDate) {
    const diff = daysBetween(habit.lastCompletedDate, today);
    if (diff === 1) {
      habit.currentStreak = (habit.currentStreak || 0) + 1;
    } else {
      habit.currentStreak = 1;
    }
  } else {
    habit.currentStreak = 1;
  }

  if (habit.currentStreak > (habit.longestStreak || 0)) {
    habit.longestStreak = habit.currentStreak;
  }

  habit.lastCompletedDate = today;
  const saved = await habitRepo.save(userId, habit);
  return toDTO(saved);
}

async function getHabitLogs(userId, habitId, startDate, endDate) {
  const habit = await habitRepo.findById(userId, habitId);
  if (!habit) throw notFound(habitId);
  const logs = await habitLogRepo.findByDateRange(userId, habitId, startDate, endDate);
  return logs.map(l => ({ date: l.date, completed: l.completed, completedAt: l.completedAt || null }));
}

async function getDashboard(userId) {
  const habits = await getHabits(userId);
  const endDate = todayStr();
  const startDate = dateStrMinusDays(30);

  const logs = {};
  for (const h of habits) {
    logs[h.id] = await getHabitLogs(userId, h.id, startDate, endDate);
  }

  const intelligence = computeIntelligence(habits, logs);
  return { habits, logs, intelligence };
}

function computeIntelligence(habits, logsMap) {
  if (!habits.length) {
    return { consistencyScore: 0, riskCount: 0, longestStreak: 0, weeklyCompletionRate: 0 };
  }

  const today = new Date();

  // 1. Weighted consistency score (30 days, recency-weighted)
  let weightedCompletion = 0;
  let totalWeight = 0;

  for (let i = 0; i < 30; i++) {
    const dateStr = dateStrMinusDays(i);
    const weight = (30 - i) / 30;
    totalWeight += weight * habits.length;

    for (const h of habits) {
      const hLogs = logsMap[h.id] || [];
      if (hLogs.some(l => l.date === dateStr && l.completed)) {
        weightedCompletion += weight;
      }
    }
  }
  const consistencyScore = totalWeight > 0 ? Math.round((weightedCompletion / totalWeight) * 100) : 0;

  // 2. Risk count — daily habits with streak but not completed today or yesterday
  const todayS = todayStr();
  const yesterdayS = dateStrMinusDays(1);
  let riskCount = 0;
  for (const h of habits) {
    if ((h.frequency === 'daily' || !h.frequency) &&
        h.currentStreak > 0 &&
        h.lastCompletedDate !== todayS &&
        h.lastCompletedDate !== yesterdayS) {
      riskCount++;
    }
  }

  // 3. Longest current streak
  const longestStreak = habits.reduce((max, h) => Math.max(max, h.currentStreak || 0), 0);

  // 4. Weekly completion rate
  let weeklyCompleted = 0;
  const weeklyExpected = habits.length * 7;
  for (let i = 0; i < 7; i++) {
    const dateStr = dateStrMinusDays(i);
    for (const h of habits) {
      const hLogs = logsMap[h.id] || [];
      if (hLogs.some(l => l.date === dateStr && l.completed)) weeklyCompleted++;
    }
  }
  const weeklyCompletionRate = weeklyExpected > 0 ? Math.round((weeklyCompleted / weeklyExpected) * 100) : 0;

  return { consistencyScore, riskCount, longestStreak, weeklyCompletionRate };
}

function toDTO(habit) {
  return {
    id: habit.id,
    name: habit.name,
    description: habit.description || null,
    category: habit.category || null,
    frequency: habit.frequency || 'daily',
    goalType: habit.goalType || null,
    goalValue: habit.goalValue || 0,
    motivation: habit.motivation || null,
    difficulty: habit.difficulty || 1,
    currentStreak: habit.currentStreak || 0,
    longestStreak: habit.longestStreak || 0,
    streakFreezeAvailable: habit.streakFreezeAvailable || 0,
    lastCompletedDate: habit.lastCompletedDate || null,
    createdAt: toISO(habit.createdAt),
  };
}

module.exports = { createHabit, getHabits, getHabit, updateHabit, deleteHabit, completeHabit, getHabitLogs, getDashboard };
